### PR TITLE
add victoria-metrics-distributed chart

### PR DESCRIPTION
**Pull request by Haleygo**
address https://github.com/VictoriaMetrics/helm-charts/issues/789

### CODE REVIEW

File: charts/victoria-metrics-distributed/README.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+* Install the follow packages: ``git``, ``kubectl``, ``helm``, ``helm-docs``. See this [tutorial](../../REQUIREMENTS.md).
+
+* PV support on underlying infrastructure.
+
+* Multiple availability zones.
+
+## Chart Details
+
+This chart is for setting up multiple VictoriaMetrics cluster instances on multiple [availability zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/).
+The default values suppose there are two availability zones with label `topology.kubernetes.io/zone: zone-a` and `topology.kubernetes.io/zone: zone-a`, default topology will be like below:
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:20]: Should it be zone-a and zone-b?

File: charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg
Comment (hagen1778) [inline charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg:1]: On the read path, why there is `vmauth-query` in zone-a and `vmauth-region-query` in zone-b? shouldn't they have the same name? wdyt about renaming them to `vmauth-read-balancer`. And on ingestion path, `vmauth-write-balancer`?

Comment (hagen1778) [reply charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg:1]: Although, why vminserts and vmselects visually have different number of replicas?

File: charts/victoria-metrics-distributed/README.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+![victoriametrics-distributed-topology](./victoriametrics-distributed-topology.jpg)
+
+It contains required availability zone components and global components, and other optional resources.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:23]: This sentence doesn't contribute additional knowledge and can be skipped.

File: charts/victoria-metrics-distributed/README.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+Availability zone components:
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:25]:
```suggestion
Availability zone components:
```
Each AZ contains the following components:

File: charts/victoria-metrics-distributed/README.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+1. vmagent: remotes write to all availability zones to ensure data completeness.
+2. vmauth-ingest: performs `least_loaded` policy for spreading requests to vminsert instances located in the same availability zone.
+3. vmcluster: supports different setup per availability zone.
+4. vmauth-query: performs `least_loaded` policy for spreading requests to vmselect instances located in the same availability zone.
+5. vmauth-query-cross-az: performs `first_available` policy on vmselect instances of all the availability zones, prefers "local" vmselect to reduce cross zone network traffic.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:30]: Can it be substituted with ` vmauth-global-query`?

File: charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg
Comment (hagen1778) [reply charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg:1]: Would be nice to plot this in excalidraw and add source code of the picture as well. Later, we might want to update the picture and it would be convenient to re-use its source code.

File: charts/victoria-metrics-distributed/README.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+1. using one vmcluster but spread the components on different zones by `topologySpreadConstraints`.
+For example, set up vmcluster with `replicas: 2` for vminsert, vmstorage and vmselect with `replicationFactor=2`, and set `topologySpreadConstraints` for all of the components:
+```
+    topologySpreadConstraints:
+      - maxSkew: 1
+        topologyKey: topology.kubernetes.io/zone
+        whenUnsatisfiable: DoNotSchedule
+```
+Each zone will have a pair of nodes of each type - vminsert, vmselect and vmstorage, thus vmcluster can still serve both write and query when one zone is compeletely down.
+But VictoriaMetrics components themselves don't support zone-awareness, means in the above example, if you have three vmstorage nodes with `replicationFactor=2`, it can't be ensure that duplicated data are written to vmstorage instances on different availability zone, thus data could be incomplete when one availability zone is down.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:54]: Doesn't this mean this approach is incorrect?) If yes, we shouldn't mention it.

File: charts/victoria-metrics-distributed/README.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+## Prerequisites
+
+* Install the follow packages: ``git``, ``kubectl``, ``helm``, ``helm-docs``. See this [tutorial](../../REQUIREMENTS.md).
+
+* PV support on underlying infrastructure.
+
+* Multiple availability zones.
+
+## Chart Details
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:17]: Thanks for writing the doc! But I find it lacking a bit of details on the topology. Maybe explaining failure scenarios and how this topology helps to provide availability will be more understandable for reader.

File: charts/victoria-metrics-distributed/README.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+## Parameters
+
+The following tables lists the configurable parameters of the chart and their default values.
+
+Change the values according to the need of the environment in ``victoria-metrics-distributed`/values.yaml`` file.
+
+| Key | Type | Default | Description |
+|-----|------|---------|-------------|
+| availabilityZones[0].allowIngest | bool | `true` |  |
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:200]: Is it planned to add description to these params at some point?

File: charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg
Comment (Haleygo) [reply charts/victoria-metrics-distributed/victoriametrics-distributed-topology.jpg:1]: all fixed, plz check

File: charts/victoria-metrics-distributed/README.md (hunk 2 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md]: removed

File: charts/victoria-metrics-distributed/README.md (hunk 3 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md]: rephrased

File: charts/victoria-metrics-distributed/README.md (hunk 4 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md]: changed some namse, plz check the latest and let me know)

File: charts/victoria-metrics-distributed/README.md (hunk 5 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md]: yes, removed

File: charts/victoria-metrics-distributed/README.md (hunk 7 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md]: they were there, but with wrong comment format, fixed.

File: charts/victoria-metrics-distributed/README.md (hunk 1 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md]: fixed!

File: charts/victoria-metrics-distributed/README.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+### How to query data?
+
+The chart provides `vmauth-global-read` as global read entrypoint, it picks the first zone as it's preferred datasource and switches automatically to next zone if first one is unavailable, check [vmauth `first_available`](https://docs.victoriametrics.com/vmauth/#high-availability) for more details.  
+If you have workloads like vmalert or grafana which deployed on one zone specifically, you can use `vmauth-read-proxy` on the same zone as your datasource, since it always prefer "local" vmcluster which reduces cross-zone traffic. 
+
+You can also pick other proxies like kubernetes service which supports [Topology Aware Routing](https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/) as global read entrypoint.
+
+### What will happen if one zone outraged?
+
+If availability zone `zone-eu-1` outraged, `vmauth-global-write` and `vmauth-global-read` still work without interruption, and following things happen:
```
Comment (zekker6) [inline charts/victoria-metrics-distributed/README.md:64]:
```suggestion
### What will happen in case of zone outage?

In case availability zone `zone-eu-1` is not available, `vmauth-global-write` and `vmauth-global-read` still work without interruption, and following things happen:
```

Comment (zekker6) [reply charts/victoria-metrics-distributed/README.md:64]: Most likely "outrage" was used due to a typo :)

File: charts/victoria-metrics-k8s-stack/tmp.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
```
Comment (zekker6) [inline charts/victoria-metrics-k8s-stack/tmp.yaml:1]: is this file needed?

**Review by zekker6 — DISMISSED** [review]
LGTM

nit: a small comment, also would be great to also add lint setup for this chart in the [Makefile](https://github.com/VictoriaMetrics/helm-charts/blob/aad26d596ed88e0bfbbe5b05a491b6e149e8b4b6/Makefile#L62-L83)

File: charts/victoria-metrics-distributed/README.md (hunk 8 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md:64]: yes, thanks!

File: charts/victoria-metrics-k8s-stack/tmp.yaml (hunk 9 above)
Comment (Haleygo) [reply charts/victoria-metrics-k8s-stack/tmp.yaml]: ofc no, some tests got mixed)

**Review by zekker6 — APPROVED** [review]

File: charts/victoria-metrics-distributed/README.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+To avoid this, vmcluster must be installed on multiple availability zones and each of them should contain complete data. In this way, as long as one zone is available, both global write and read entrypoint work without interruption.
+
+### How to write data?
+
+The chart provides `vmauth-global-write` as global write entrypoint, it supports [push-based data ingestion protocols](https://docs.victoriametrics.com/vmagent/#how-to-push-data-to-vmagent) as VictoriaMetrics does.
+You can use this entrypoint as vmagent's remoteWrite url or use other clients push data to it directly.
+
+### How to query data?
+
+The chart provides `vmauth-global-read` as global read entrypoint, it picks the first zone as it's preferred datasource and switches automatically to next zone if first one is unavailable, check [vmauth `first_available`](https://docs.victoriametrics.com/vmauth/#high-availability) for more details.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:57]:
```suggestion
The chart provides `vmauth-global-read` as global read entrypoint, it picks the first available zone (see [first_available](https://docs.victoriametrics.com/vmauth/#high-availability) policy) as it's preferred datasource and switches automatically to next zone if first one is unavailable.  
```

File: charts/victoria-metrics-distributed/README.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+### How to write data?
+
+The chart provides `vmauth-global-write` as global write entrypoint, it supports [push-based data ingestion protocols](https://docs.victoriametrics.com/vmagent/#how-to-push-data-to-vmagent) as VictoriaMetrics does.
+You can use this entrypoint as vmagent's remoteWrite url or use other clients push data to it directly.
+
+### How to query data?
+
+The chart provides `vmauth-global-read` as global read entrypoint, it picks the first zone as it's preferred datasource and switches automatically to next zone if first one is unavailable, check [vmauth `first_available`](https://docs.victoriametrics.com/vmauth/#high-availability) for more details.  
+If you have workloads like vmalert or grafana which deployed on one zone specifically, you can use `vmauth-read-proxy` on the same zone as your datasource, since it always prefer "local" vmcluster which reduces cross-zone traffic.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:58]:
```suggestion
If you have services like [vmalert](https://docs.victoriametrics.com/vmalert.html) or Grafana deployed in each zone, then configure them to use local `vmauth-read-proxy`. Per-zone `vmauth-read-proxy` always prefers "local" vmcluster for querying and reduces cross-zone traffic. 
```

File: charts/victoria-metrics-distributed/README.md [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+If you have workloads like vmalert or grafana which deployed on one zone specifically, you can use `vmauth-read-proxy` on the same zone as your datasource, since it always prefer "local" vmcluster which reduces cross-zone traffic. 
+
+You can also pick other proxies like kubernetes service which supports [Topology Aware Routing](https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/) as global read entrypoint.
+
+### What happens if zone outage happened?
+
+If availability zone `zone-eu-1` had an outage, `vmauth-global-write` and `vmauth-global-read` still work without interruption, and following things happen:
+1. `vmauth-global-write` stops proxying write requests to `zone-eu-1` automatically;
+2. `vmauth-global-read` and `vmauth-read-proxy` stops proxying read requests to `zone-eu-1` automatically;
+3. `vmagent` on `zone-us-1` fails to send data to `zone-eu-1` `vmauth-write-balancer`, starts to [buffer data on disk](https://docs.victoriametrics.com/vmagent/#calculating-disk-space-for-persistence-queue) if `-remoteWrite.disableOnDiskQueue` is not specified.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:67]:
```suggestion
3. `vmagent` on `zone-us-1` fails to send data to `zone-eu-1.vmauth-write-balancer`, starts to [buffer data on disk](https://docs.victoriametrics.com/vmagent/#calculating-disk-space-for-persistence-queue) (unless `-remoteWrite.disableOnDiskQueue` is specified, which is not recommended for this topology).
```

File: charts/victoria-metrics-distributed/README.md [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:68]: Once the availability of `zone-eu-1` is restored, `vmagent` from `zone-us-1` will re-send all the buffered data to `zone-eu-1` to restore data consistency.

Comment (hagen1778) [reply charts/victoria-metrics-distributed/README.md:68]: @Haleygo @zekker6 when will vmauth start to route requests to `zone-eu-1`? Could it be that after the outage vmauth will route queries to `zone-eu-1` and will serve incomplete responses?

File: charts/victoria-metrics-distributed/README.md [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+This chart sets up multiple VictoriaMetrics cluster instances on multiple [availability zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/), provides both global write and read entrypoints.
+
+The default setup is as shown below:
+
+![victoriametrics-distributed-topology](./victoriametrics-distributed-topology.png)
+
+For write:
+1. extra-vmagent(optional): scrapes external targets and all the components installed by this chart, sends data to global write entrypoint.
+2. vmauth-global-write: global write entrypoint, proxies requests to one of the zone `vmagent` with `least_loaded` policy.
+3. vmagent(per-zone): remote writes data to availability zones that enabled `.Values.availabilityZones.allowIngest`.
```
Comment (hagen1778) [inline charts/victoria-metrics-distributed/README.md:28]: we need to mention that it acts as a buffer in the same time, for restoring data consistency in case if any zone is unavailable. Hence, we need to provide recommendations regarding disk space capacity.

File: charts/victoria-metrics-distributed/README.md (hunk 13 above)
Comment (zekker6) [reply charts/victoria-metrics-distributed/README.md:68]: Based on [this](https://github.com/VictoriaMetrics/helm-charts/blob/64b4f71f94139548de4147c06a0907983901aaab/charts/victoria-metrics-distributed/victoriametrics-distributed-topology.png) diagram it will start routing requests after vmauth read proxy will become available. So in this case it is possible that it will deliver incomplete responses.

Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md:71]: Add more description for it, plz check https://github.com/VictoriaMetrics/helm-charts/pull/960/commits/19f29bedc9a77a5c285cc84bfdca32ef3f8ac80c

File: charts/victoria-metrics-distributed/README.md [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+## Prerequisites
+
+* Install the follow packages: ``git``, ``kubectl``, ``helm``, ``helm-docs``. See this [tutorial](../../REQUIREMENTS.md).
+
+* PV support on underlying infrastructure.
+
+* Multiple availability zones.
+
+## Chart Details
```
Comment (Amper) [inline charts/victoria-metrics-distributed/README.md:17]: 1. I think we should place scheme from issue here
2. Maybe we should describe upgrade scenario for the cluster step-by-step

File: charts/victoria-metrics-distributed/values.yaml [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+availabilityZones:
+  - name: zone-eu-1
+    # -- allow data ingestion to this zone
+    allowIngest: true
+    # -- allow data query from this zone through global query endpoint
+    allowQuery: true
+    # -- nodeselector to restrict where pods of this zone can be placed.
+    # usually provided by cloud providers.
+    nodeSelector:
+      topology.kubernetes.io/zone: zone-eu-1
```
Comment (Amper) [inline charts/victoria-metrics-distributed/values.yaml:29]: The scenario of using `nodeAffinity` instead of `nodeSelector` for zone assignment is not taken into account.

File: charts/victoria-metrics-distributed/README.md (hunk 15 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/README.md:17]: added in https://github.com/VictoriaMetrics/helm-charts/pull/960/commits/4e3204ed23a119eb48b9a2eb6d9a8ce8f84156a8, please check.

File: charts/victoria-metrics-distributed/values.yaml (hunk 16 above)
Comment (Haleygo) [reply charts/victoria-metrics-distributed/values.yaml:29]: added in https://github.com/VictoriaMetrics/helm-charts/pull/960/commits/4e3204ed23a119eb48b9a2eb6d9a8ce8f84156a8, please check.

**Review by hagen1778 — APPROVED** [review]
LGTM!
But I'd prefer to see how it works in practice in sandbox ENV. Would it be possible to demonstrate, @Haleygo ?

**Review by Amper — APPROVED** [review]
LGTM

**Comment by Haleygo** [comment]
@hagen1778 @Amper please check one more fix https://github.com/VictoriaMetrics/helm-charts/pull/960/commits/16705166a0a61b4fca8fb7eff1c3d736e48ef08f

**Review by Amper — DISMISSED** [review]

**Review by zekker6 — APPROVED** [review]

**Review by zekker6 — APPROVED** [review]
